### PR TITLE
C++: Fix missing results for `comparesEq` in `IRGuardCondition`

### PR DESCRIPTION
**Pull request by MathiasVP**
Discovered while working on https://github.com/github/codeql/pull/16677. After this PR has been merged we can get rid of the spurious results added in https://github.com/github/codeql/pull/16677.

Commit-by-commit recommended.

Edit: https://github.com/github/codeql/pull/16677 has now been merged into `main`, so I've merged `main` into this branch and updated the test to reflect the impact of this PR

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 1]
```diff
@@ -856,11 +856,15 @@ private predicate unary_simple_comparison_eq(
     inNonZeroCase = false
   )
   or
-  // There's no implicit CompareInstruction in files compiled as C since C
-  // doesn't have implicit boolean conversions. So instead we check whether
-  // there's a branch on a value of pointer or integer type.
-  relevantUnaryComparison(test) and
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:861]: That whole predicate can be removed now, as the only remaining use is a recursion.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:861]: Oh, right. Well spotted! Fixed in 7f62085be57ef80678323c1de5e964ea6a5afbed

**Comment by MathiasVP** [comment]
> DCA highlights some stage times that got slower. Do those need to be investigated?

Indeed, it does look like the stage that computes `IRGuard`s became slower. However, it doesn't look like it's reflected in the actual analysis time. So I don't think this is a real issue that needs to be investigated, no.

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 2]
```diff
@@ -856,11 +856,15 @@ private predicate unary_simple_comparison_eq(
...
-  op.getDef() = test and
+  not test.isGLValue() and
+  not simple_comparison_eq(test, _, _, _, _) and
+  not simple_comparison_lt(test, _, _, _) and
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:842]: The exclusion of switch expressions I understand, as these are handled above, but why are these three lines the correct thing to do?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:842]: It doesn't really matter. Without these conjuncts, this predicate basically consists of all instructions of type address, integer, or boolean. This is simply a very easy way of reducing this relatively large set.

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:842]: Thanks.

> by excluding instructions which are already guard conditions for other reasons (i.e, by being included in `simple_comparison_eq` or `simple_comparison_lt `), or being irrelevant (i.e., by being a glvalue).

This is indeed what I expect. However, I have a hard time following the logic of the library here. For example, I started looking at the places where `unary_simple_comparison_eq` is used, and expected `simple_comparison_eq` or `simple_comparison_lt` or `isGLValue` there, but didn't see those predicates mentioned.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:842]: Yeah, I can see why this is hard to follow. `simple_comparison_eq` is used in another code path (when we infer `e1 == e2 + k` equalities) in [compares_eq](https://github.com/github/codeql/blob/06aa2664bf1ec92c43751d73a215d2d26b65b640/cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll#L716) (which is used when a query uses [GuardCondition::comparesEq](https://github.com/github/codeql/blob/06aa2664bf1ec92c43751d73a215d2d26b65b640/cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll#L558)), and `simple_comparison_lt` is used in _yet another_ code path (where we infer `e1 < e2 + k` inequalities) when a query uses [GuardCondition::comparesLt](https://github.com/github/codeql/blob/06aa2664bf1ec92c43751d73a215d2d26b65b640/cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll#L492).

I'll write up a comment explaining this in this case. And now that I'm here I'll write some more general comments in this file as well.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:895]: I've added some explanatory comments to 7819cc1c36fda57fdac1421f9cf8b32c570f366e. Does that help in this case?

**Review by jketema — APPROVED** [review]
LGTM

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 2 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:895]: Thanks. That helps a lot.
